README.txt
ZSQL - Zope Python product and Z SQL Method bridge
==================================================

Overview
--------

 License: ZPL
 Copyright, Peter Bengtsson, peter@fry-it.com, Fry-IT, 2007

 This product is meant to be used my other Zope Python products where 
 you want to have the SQL code on the filesystem as a .sql files and
 use DTML for the rendering. 
 This product helps setting up the Z SQL Methods seemlessly with some
 added debugging and optimization features.
 
 
Usage
-----

 You write a little wrapper in your product and define the classes of
 Z SQL Methods you want to use and put the .sql files inside the product
 with directory names like you please. Assuming you have these file:
 
 __init__.py
 MyApp.py
 SQLWrapper.py
 sql/
   Users/
     SQLSelectUsers.sql
     SQLInsertUser.sql
   Cookies/
     SQLSelectCookies.sql
     SQLDeleteCookie.sql
  
 Then, inside SQLWrapper.py you put the following code::
 
  import Acquisition
  from Globals import package_home
  from Products.ZSQL import initializeSQLfiles
  
  sqlhome = os.path.join(package_home(globals()), 'sql')
  
  class PSQL(Acquisition.Implicit):
      dbConnection = 'Psycopg_database_connection'
  class SQLUsers(PSQL):
      pass
  class SQLCookies(PSQL):
      pass

  Folder2Class = {
    'Cookies':SQLCookies,
    'Users':SQLUsers,
    }
    
  initializeSQLfiles(Folder2Class, sqlhome)
  
  
 Once you've done this, the classes 'SQLUsers' and 'SQLCookies' will
 have automatically gotten the Z SQL Methods 'SQLSelectUsers',
 'SQLInsertUser' and 'SQLSelectCookies', 'SQLDeleteCookie'.

 Now you can use these are normal member functions in your main class.
 Inside MyApp.py::
 
  class MyApp(Folder):
      meta_type = 'Something'
      def __init__(self, id):
          self.id = id
          
      def addUser(self, name):
          for user in self.SQLSelectUsers():
              print user.name
          self.SQLInsertUser(name=name)
          
      def removeCookie(self, key):
          self.SQLDeleteCookie(key=key)
          
          
 The .sql files **have** to start with <params>...</params>. For
 example, it can look like this::
 
  <params>
  key
  name
  </params>
  INSERT INTO users (key, name) VALUES (
  <dtml-sqlvar key type="int">,  <dtml-sqlvar name type="string">
  )
  
  
 
Features
--------

 One useful feature is that you can easily debug the rendered SQL code
 and have it printed to stdout by calling it like this::
 
  SQLInsertUser(key=123, name='Peter', debug__=True)
  
 Or if you want to use memcache for those slow database queries that
 you repeat like this::
 
  SQLSelectAverageUserAge(min=10, max=20, memcache__=True)
  
 If you call this a second time (within 5 minutes) the result will
 come from a memcache instead. This only becomes useful if the SQL
 query is slow and for this to work you need to have a memcache server
 running on 127.0.0.1:11211.
 
 If you want to debug all SQL calls to stdout you can call 
 the initializeSQLfiles() method like this instead::
 
  initializeSQLfiles(Folder2Class, sqlhome, debug_sqlcalls=True)
  
 If you also want to have all of these written to file you can
 do that this way::
 
  initializeSQLfiles(Folder2Class, sqlhome, debug_sqlcalls=True,
                     debug_sqlcalls_filename='/tmp/sqlcalls.log')
                      
 Similar to the 'debug_sqlcalls' and 'debug_sqlcalls_filename' is
 the profiling option except that when profiling is on, it's not
 printed to stdout. If you set up your SQL files like this::
 
  initializeSQLfiles(Folder2Class, sqlhome, profile_sqlcalls=True)
  
 Then it writes all profiling records to /tmp/sqlprofiling.log.
 Alternatively you can override this like this::
 
  initializeSQLfiles(Folder2Class, sqlhome, profile_sqlcalls=True,
          profile_sqlcalls_filename=r'C:\Temp\sqlprofileing.log')
          
       
 
 
  
 


CHANGES.txt
-

  * You can now pass explain__=True to any call and you get the SQL
    run with 'EXPLAIN ANALYZE ' prefixed. (PostgreSQL only)

  * Debug printing queries that contain Unicode could cause a
    UnicodeDecodeError when printing to stdout.

- 0.9.3

  * If the python memcache module wasn't installed and the call wants
    to use it it wasn't caught in a warning message.

- 0.9.2

  * Fixed a bug in how debug() was called with a parameter 'f'.
  
  * Made key be generated from parameter values instead of keys and
    values.
    
- 0.9.1

  * Added warning when memcache module not available but is trying to
    be used.
    
- 0.9
 
  * Started keeping a CHANGES.txt
  
 

tests/testSQL.py
# -*- coding: iso-8859-1 -*
# ZSQL - Zope Python product and Z SQL Method bridge
# License: ZPL
# Copyright: Peter Bengtsson, Fry-IT, 2007
#

import os, sys, re
import cPickle as pickle

if __name__ == '__main__':
    execfile(os.path.join(sys.path[0], 'framework.py'))


from Testing import ZopeTestCase

from Products.ZSQL.SQL import SimpleRecord, MemCache, InterceptedSQLClass

class TestSQL(ZopeTestCase.ZopeTestCase):
    
    def afterSetUp(self):
        pass
    
    def tearDown(self):
        pass
    
    
    def test_SimpleRecord_pickle(self):
        """ test that we can pickle (protocol 2) SimpleRecord objects """
        friend = SimpleRecord(name='Zahid', age=40)
        as_pickle = pickle.dumps(friend, 2)
        as_object = pickle.loads(as_pickle)
        self.assertEqual(as_object.keys(), friend.keys())
        self.assertEqual(as_object.values(), friend.values())
        
        
    def test_MemCache(self):
        cache = MemCache(('127.0.0.1:11211',))
        try:
            assert 1 == cache.set('os.listdir("/tmp")', os.listdir("/tmp"))
        except AssertionError:
            class MemcacheError(Exception):
                pass
            raise MemcacheError("memcache server not running")
        
        self.assertEqual(cache.get('os.listdir("/tmp")'), os.listdir("/tmp"))
        self.assertEqual(cache.get('not-set-yet'), None)

        cache.delete('os.listdir("/tmp")')
        self.assertEqual(cache.get('os.listdir("/tmp")'), None)
        
        # test the timeout
        cache.set('os.listdir("/tmp")', os.listdir("/tmp"), timeout=1)
        self.assertEqual(cache.get('os.listdir("/tmp")'), os.listdir("/tmp"))
        from time import sleep
        sleep(2)
        self.assertEqual(cache.get('os.listdir("/tmp")'), None)
        
        
        # an incorrect port number, don't expect a memcache server to run there
        cache = MemCache(('127.0.0.1:11212',))
        self.assertEqual(cache.set('os.listdir("/tmp")', os.listdir("/tmp")), 0)
        
        
    def test_InterceptedSQLClass(self):
        klass = InterceptedSQLClass('SQLSelect', 'Somet title', 'db-connection',
                                    'arg1 arg2', 'select * from users',
                                    'SQLselect.sql')
        
        self.assertEqual(klass.getRelpath(), 'SQLselect.sql')
        
#
# Need to write more tests
#
        


def test_suite():
    from unittest import TestSuite, makeSuite
    suite = TestSuite()
    suite.addTest(makeSuite(TestSQL))
    return suite

if __name__ == '__main__':
    framework()

    


tests/framework.py
##############################################################################
#
# Copyright (c) 2005 Zope Corporation and Contributors. All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""ZopeTestCase framework

COPY THIS FILE TO YOUR 'tests' DIRECTORY.

This version of framework.py will use the SOFTWARE_HOME
environment variable to locate Zope and the Testing package.

If the tests are run in an INSTANCE_HOME installation of Zope,
Products.__path__ and sys.path with be adjusted to include the
instance's Products and lib/python directories respectively.

If you explicitly set INSTANCE_HOME prior to running the tests,
auto-detection is disabled and the specified path will be used 
instead.

If the 'tests' directory contains a custom_zodb.py file, INSTANCE_HOME
will be adjusted to use it.

If you set the ZEO_INSTANCE_HOME environment variable a ZEO setup 
is assumed, and you can attach to a running ZEO server (via the 
instance's custom_zodb.py).

The following code should be at the top of every test module:

  import os, sys
  if __name__ == '__main__':
      execfile(os.path.join(sys.path[0], 'framework.py'))

...and the following at the bottom:

  if __name__ == '__main__':
      framework()

$Id: framework.py 12108 2007-04-01 18:02:38Z peterbe $
"""

__version__ = '0.2.4'

# Save start state
#
__SOFTWARE_HOME = os.environ.get('SOFTWARE_HOME', '')
__INSTANCE_HOME = os.environ.get('INSTANCE_HOME', '')

if __SOFTWARE_HOME.endswith(os.sep):
    __SOFTWARE_HOME = os.path.dirname(__SOFTWARE_HOME)

if __INSTANCE_HOME.endswith(os.sep):
    __INSTANCE_HOME = os.path.dirname(__INSTANCE_HOME)

# Find and import the Testing package
#
if not sys.modules.has_key('Testing'):
    p0 = sys.path[0]
    if p0 and __name__ == '__main__':
        os.chdir(p0)
        p0 = ''
    s = __SOFTWARE_HOME
    p = d = s and s or os.getcwd()
    while d:
        if os.path.isdir(os.path.join(p, 'Testing')):
            zope_home = os.path.dirname(os.path.dirname(p))
            sys.path[:1] = [p0, p, zope_home]
            break
        p, d = s and ('','') or os.path.split(p)
    else:
        print 'Unable to locate Testing package.',
        print 'You might need to set SOFTWARE_HOME.'
        sys.exit(1)

import Testing, unittest
execfile(os.path.join(os.path.dirname(Testing.__file__), 'common.py'))

# Include ZopeTestCase support
#
if 1:   # Create a new scope

    p = os.path.join(os.path.dirname(Testing.__file__), 'ZopeTestCase')

    if not os.path.isdir(p):
        print 'Unable to locate ZopeTestCase package.',
        print 'You might need to install ZopeTestCase.'
        sys.exit(1)

    ztc_common = 'ztc_common.py'
    ztc_common_global = os.path.join(p, ztc_common)

    f = 0
    if os.path.exists(ztc_common_global):
        execfile(ztc_common_global)
        f = 1
    if os.path.exists(ztc_common):
        execfile(ztc_common)
        f = 1

    if not f:
        print 'Unable to locate %s.' % ztc_common
        sys.exit(1)

# Debug
#
print 'SOFTWARE_HOME: %s' % os.environ.get('SOFTWARE_HOME', 'Not set')
print 'INSTANCE_HOME: %s' % os.environ.get('INSTANCE_HOME', 'Not set')
sys.stdout.flush()



tests/run_testSQL.sh
#!/bin/sh
export SOFTWARE_HOME=/home/peterbe/zope/zope288/lib/python/
export INSTANCE_HOME=/home/peterbe/zope/zope288/
python2.4 testSQL.py $@



SQL.py
# -*- coding: iso-8859-1 -*
# ZSQL - Zope Python product and Z SQL Method bridge
# License: ZPL
# Copyright, Peter Bengtsson, Fry-IT, 2007
# http://www.fry-it.com/oss/ZSQL

# Python
import os, time, sys
import inspect
import logging

# Zope
from AccessControl import ClassSecurityInfo
from Globals import InitializeClass
from Globals import package_home
from Products.ZSQLMethods.SQL import SQL
from DateTime import DateTime

from App.config import getConfiguration
conf = getConfiguration()
ZOPE_DEBUG_MODE = conf.debug_mode

# Product
class MissingArgumentsError(Exception):
    pass
class ExcessArgumentsError(Exception):
    pass

def debug(s):
    print s


try:
    import memcache as memcache_
    __memcache_installed__ = True
except ImportError:
    memcache_ = None
    __memcache_installed__ = False


logger = logging.getLogger('ZSQL')

# I've done some testing with cmemcache instead of memcache and found that the
# difference is also negligable. Ie. using cmemcache instead of python-memcache
# doesn't make it any faster. python-memcache is fast enough. What takes time
# isn't that, it's the other stuff around it. For example, I did a test on
# getting 500 * 10 reads first with python-memcache and then with cmemcache.
# With python-memcache it took 18.7 seconds, with cmemcache it took 18.1 :)
#import cmemcache as memcache




##
## Possible TODO, have a look what the fastest template rendering
## engine is to be used instead of DTML. Perhaps DTML is very fast
## and the cost of replacing it with something like WebString might
## have dire consequences on the complete rendering of the ZSQL
## method.
##


def _debugSQLCall(object, kw, debug_filename):
    """ Write to debug file for log of all SQL calls """
    if not debug_filename:
        debug_filename = '/tmp/sqllog.log'
    path = os.path.join(CLIENT_HOME, debug_filename)
    
    sql_method = str(object.id)
    params = str(kw)
    relpath = object.relpath
    
    datetime = DateTime().strftime('%Y/%m/%d %H:%M:%S')
    line = '%s|%s|%s|%s\n'%(sql_method, relpath, params, datetime)
    open(path, 'a').write(line)
    
    
def _profileSQLCall(object, timetaken, kw, profiling_filename):
    """ write to log file how long it took to execute
    this SQL statement. """
    if not profiling_filename:
        profiling_filename = '/tmp/sqlprofiling.log'
    path = os.path.join(CLIENT_HOME, profiling_filename)
    fw = open(path, 'a')
        
    out = [str(object.relpath), str(timetaken)]
    out = '|'.join(out)
    
    fw.write(out+'\n')
    fw.close()

    
    
class SimpleRecord(dict):
    """ Convert a Zope ZSQL Record into a dict like object that works 
    like a dict but with a __getattr__ method.
    """
    __allow_access_to_unprotected_subobjects__ = 1
    
    def __init__(self, *a, **kw):
        if a:
            if hasattr(a[0], '__record_schema__'):
                for k, v in a[0].__record_schema__.items():
                    kw[k] = a[0][v]
                    
            elif isinstance(a[0], dict):
                kw.update(a[0])
        dict.__init__(self, **kw)
        
    def __getattr__(self, key):
        """ see 
        http://groups.google.com/group/comp.lang.python/browse_thread/thread/40d95aa4a99f3b87/7219e0c75af69198 """
        try:
            return self[key]
        except KeyError:
            raise AttributeError 



class MemCache(object):
    def __init__(self, servers):
        self._cache = memcache_.Client(servers)
        self.default_timeout = 300

    def get(self, key, default=None):
        val = self._cache.get(key)
        if val is None:
            return default
        else:
            return val

    def set(self, key, value, timeout=0):
        return self._cache.set(key, value, timeout or self.default_timeout)

    def delete(self, key):
        return self._cache.delete(key)

    def get_many(self, keys):
        return self._cache.get_multi(keys)
    
    
        
class InterceptedSQLClass(SQL):
    """ subclass of the SQL (from ZSQLMethods) so that
    we can enable possible executions and initializations.
    
    
    @ memcache_prefix : A string that prepends on all memcaching keys. Useful if you 
                       fear that your application might share the same path and params
                       as in another application.
    @ memcache_servers : a list or tuple of addresses to memcache servers
    @ debug_sqlcalls : All rendered SQL is printed and written to a log file
    @ debug_sqlcalls_filename : If you don't want to log all rendered SQL in
                                /tmp/sqllog.log
    @ profile_sqlcalls : Take the time it takes to run the SQL command.
    @ profile_sqlcalls_filename : If you don't want to log all profiling 
                                  in /tmp/sqlprofiling.log
                                
    """
    
    def __init__(self, id, title, connection_id, arguments, template, relpath, 
                 memcache_prefix='', 
                 memcache_servers=('127.0.0.1:11211',),
                 debug_sqlcalls=False, 
                 debug_sqlcalls_filename=None,
                 profile_sqlcalls=False,
                 profile_sqlcalls_filename=None):
        self.id=str(id)
        try:
            self.manage_edit(title, connection_id, arguments, template)
        except:
            debug("id=%r, title=%r, arguments=%r" % (id, title, arguments))
            raise 
        self.relpath = relpath
        assert isinstance(memcache_prefix, str), "memcache_prefix must be a string"
        self.memcache_prefix = memcache_prefix
        self.memcache_servers = memcache_servers
        self.debug_sqlcalls = debug_sqlcalls
        self.debug_sqlcalls_filename = debug_sqlcalls_filename
        self.profile_sqlcalls = profile_sqlcalls
        self.profile_sqlcalls_filename = profile_sqlcalls_filename


    def _initMemcache(self):
        self._v_memcache = MemCache(self.memcache_servers)
        
    
    def __call__(self, REQUEST=None, __ick__=None, src__=False, test__=False,
                 debug__=False, memcache__=False, explain__=False, **kw):
        """ override __call__ for debugging purposes
        
        @ src__ = True: return the computed query before being tested about the database
        @ test__= True: return (query, result) and not just result
        @ debug__= True: print the SQL statement before running it (if in DEBUG mode)
        @ memcache__ = True: memcaches the result for consequtive calls for 5 minutes
        @ memcache__ = int: memcaches the result for consequtive calls for 'int' seconds
        """
        
        # legacy 
        if kw.has_key('force_debug__'):
            debug__ = kw.get('force_debug__')
            kw.pop('force_debug__')
            import warnings
            warnings.warn("Use 'debug__' instead of 'force_debug__'", DeprecationWarning, 2)
        
            
        if self.debug_sqlcalls or debug__:
            if self.debug_sqlcalls_filename:
                _debugSQLCall(self, kw, self.debug_sqlcalls_filename)
                
            query = apply(SQL.__call__, (self, REQUEST, __ick__, 1, test__), kw)
            out = "--- %s(%s) ---" % (self.__name__, kw)
            
            if len(out) < 80:
                out += (80 - len(out)) * '-'
            sys.stdout.write(out + '\n')
            try:
                query = str(query)
            except UnicodeEncodeError:
                query = query.encode('ascii', 'replace')
            sys.stdout.write(query + '\n\n')
            
        if explain__:
            title, connection_id, arguments, prev_template = self.title, self.connection_id, self.arguments_src, self.src
            template = 'EXPLAIN ANALYZE\n' + prev_template
            self.manage_edit(title, connection_id, arguments, template)
            for line in apply(SQL.__call__, (self, REQUEST, __ick__, False, test__), kw):
                sys.stdout.write('%s\n' % line['QUERY PLAN'])
            self.manage_edit(title, connection_id, arguments, prev_template)
            

        if ZOPE_DEBUG_MODE:
            # If we're running Zope in debug mode, make some extra checks that
            # the parameters defined for this SQL call are correctly set.
            
            required_arguments = self._arg.keys()
            missing_arguments = [a for a in required_arguments if not kw.has_key(a)]
            excess_arguments = [k for k in kw.keys() if k not in required_arguments]
            if excess_arguments:
                tmpl = self.getRelpath()
                m = "%s was called with " % tmpl
                if len(excess_arguments) > 1:
                    m += "arguments: %s passed " %\
                         ', '.join([repr(x) for x in excess_arguments])
                else:
                    m += "argument: %r passed " % excess_arguments[0]
                m += "but not needed."
                try:
                    stk = inspect.stack()[1]
                    module = stk[1].replace(INSTANCE_HOME, '')
                    if module.startswith('/'):
                        module = module[1:]
                    module = module.replace('/','.')
                    m += "Called from %s, line %s, in %s" % (module, stk[2], stk[3])
                except IndexError:
                    m += "**Unabled to trace inspect stack**"
                raise ExcessArgumentsError, m
            
            if missing_arguments:
                tmpl = self.getRelpath()
                m = "%s was called without " % tmpl
                if len(missing_arguments) > 1: 
                    m += "arguments: %s passed." % ', '.join([repr(x) for x in missing_arguments])
                else:
                    m += "argument: %r passed." % missing_arguments[0]
                stk = inspect.stack()[1]
                module = stk[1].replace(INSTANCE_HOME, '')
                if module.startswith('/'):
                    module = module[1:]
                module = module.replace('/','.')
                m += "Called from %s, line %s, in %s" % (module, stk[2], stk[3])
                raise MissingArgumentsError, m


        if self.profile_sqlcalls:
            result, time_delta = self._getResultRecords(SQL.__call__,
                                      (self, REQUEST, __ick__, src__, test__), 
                                      kw,
                                      memcache=memcache__,
                                      profile=True)
            _profileSQLCall(self, time_delta, kw, self.profile_sqlcalls_filename)
            return result

        result, __ = self._getResultRecords(SQL.__call__, 
                                       (self, REQUEST, __ick__, src__, test__),
                                       kw, memcache=memcache__)
        return result
        
    def _getResultRecords(self, func, params, kwargs, memcache=False, profile=False):
        """ return the (records, time taken) """
        if memcache and not __memcache_installed__:
            logger.warn("Can't use memcache when memcache module is not installed")
            memcache = False
            
        if memcache:
            # Ok. We're going to try to use memcache to find the result first.
            # First make sure the non-persistent volatile memcache memory is set up.
            if not hasattr(self, '_v_memcache'):
                self._initMemcache()
                
            # Then construct a "unique" identifing key for this particular call.
            # This is constructed by combining the name of the .sql files, the keywords
            # passed and an optional prefix.
            key = self._getMemcacheKey(kwargs)
            if not hasattr(self, '_v_seen_keys'):
                self._v_seen_keys = []
            
            # Read it from the memcache. Either we get the records (nb: simplified)
            # or we get None.
            records = self._v_memcache.get(key)
            #if records is not None: 
            #    print "WOW! " * 10
                
            # Assume that it takes 0 seconds to read it from the memcache :)
            time_delta = 0
            
            # Ok. The memcache didn't have it. Let's really run the SQL command
            # and if 'profile' was passed as true, take the time it takes to
            # run the SQL command.
            if records is None:
                
                if profile:
                    # Ok. We need to wrap the actual getting in a timer
                    t0 = time.time()
                    records = apply(func, params, kwargs)
                    time_delta = time.time() - t0
                else:
                    records = apply(func, params, kwargs)
                    
                # Now set the result in the memcache for consequtive calls.
                # If the memcache passed was an integer, it will indicate how many
                # seconds are supposed to memcache it for.
                if memcache and not __memcache_installed__:
                    logger.warn("Can't use memcache when memcache module is not installed")
                    memcache = False
                    
                if memcache > 1:
                    self._v_memcache.set(key, [SimpleRecord(record) for record in records],
                                         timeout=memcache)
                else:
                    self._v_memcache.set(key, [SimpleRecord(record) for record in records])

            # Always return the records followed by the time_delta
            return records, time_delta
        else:
            # No memcache, just normal call profiled or not
            if profile:
                t0=time.time()
                records = apply(func, params, kwargs)
                return records, time.time() - t0
            else:
                return apply(func, params, kwargs), 0
            
        
    def _getMemcacheKey(self, kwargs):
        """ Return a ascii string of the relpath, keyword arguments and an
        optional prefix. Helps with the interfacing with the memcache. 
        This """
        # To make sure the key is as short as possible, remove all junk 
        relpath = self.getRelpath().replace('.sql','')
        kwargs = str(kwargs.values()).replace(' ','')
        s=[relpath, kwargs]
        if self.memcache_prefix:
            s.append(self.memcache_prefix)
        return '|'.join(s)        


    def getRelpath(self):
        return self.relpath


    
##
## Load in all the SQL statements from the files in sql/
##


def _getSQLandParams(filename):
    """ scan a SQL file and return the SQL statement as string
    and the params as a list """
    data = open(filename, 'r').read()
    statement = data[data.find('</params>')+len('</params>'):]
    paramsstr = data[data.find('<params>')+len('<params>'):data.find('</params>')]
    params = paramsstr.replace('\n',' ')
    params = params.split(' ')
    if params==['']:
        params = []
    else:
        params = [x for x in params if x]
    return params, statement.strip()




def _filterSQLextension(filenames):
    return [x for x in filenames 
            if x.lower().endswith('.sql') and not x.startswith('.#')]


def initializeSQLfiles(folder2class, sqlhomepath, **kw):
    ''' given a dictionary of foldernames and classes, return them as 
    a list of ZSQL Methods.
    For the documentation of the '**kw' options, see the doc string of
    InterceptedSQLClass.
    '''
    
    if kw.get('profile_sqlcalls'):
        log_filename = kw.get('profile_sqlcalls_filename',
                              '/tmp/sqlprofiling.log')
        logging.info("Note! All SQL calls are profiled in %s" % log_filename)
    
    for folder, Class in folder2class.items():
        Class.allsqlattributes = []
        if folder.count(':'):
            folder = folder.split(':')
            foldername = apply(os.path.join, folder)
        else:
            foldername = folder
        folder = os.path.join(sqlhomepath, foldername)
        for sqlfile in _filterSQLextension(os.listdir(folder)):
            # from the file, get the params and the statement
            params, statement = _getSQLandParams(os.path.join(folder, sqlfile))
                
            # make up an id
            id = sqlfile[:-4] # removes '.sql'

            # determine relpath
            relpath = os.path.join(folder, sqlfile).replace(sqlhomepath,'')
            
            # make ['par1:int', 'par2'] => 'par1:int par2'
            params = ' '.join(params)
            title = sqlfile
            
            # Now, create this attribute to 'Class'
            dbConnection = Class.dbConnection
            sqlclass = InterceptedSQLClass
            try:
                setattr(Class, id, sqlclass(id, title, dbConnection, 
                                            params, statement, relpath, **kw))
            except:
                print >>sys.stderr, "Failed to initialize %s" % relpath
                raise
        
    

    
            
    

